File: lint-arwaky/archive/capabilities/report_formatter_processor.py
"""report_formatter_processor — Capability for formatting reports (SARIF, JUnit)."""

from typing import Any
from ..taxonomy import (
    FileFormat,
    FORMAT_TEXT,
    FORMAT_JSON,
    FORMAT_SARIF,
    FORMAT_JUNIT,
    LogOutput,
    GovernanceReport,
    ResponseData,
)

import json
import html


from ..contract import ILintReportFormatterProtocol


class ReportFormatterProcessor(ILintReportFormatterProtocol):
    """Business logic for transforming GovernanceReports into standard formats."""

    def format(self, report: GovernanceReport) -> LogOutput:
        """Standard entry point for formatting as JSON/Dict."""
        return LogOutput(value=json.dumps(self.report_to_dict(report), indent=2))

    def get_formatted_payload(
        self, report: GovernanceReport, output_format: FileFormat = FORMAT_TEXT
    ) -> ResponseData:
        """Unified entry point for getting a formatted payload for the Surface layer."""
        data = self.report_to_dict(report)

        if output_format == FORMAT_SARIF:
            return ResponseData(
                value={"format": "sarif", "data": self.to_sarif(report).value}
            )
        if output_format == FORMAT_JUNIT:
            return ResponseData(
                value={"format": "junit", "data": self.to_junit(report).value}
            )
        if output_format == FORMAT_JSON:
            return ResponseData(value={"format": "json", "data": data})

        # Default to text/dict representation
        return ResponseData(value={"format": "text", "data": data})

    def _get_severity(self, sev: str) -> str:
        mapping = {"high": "error", "medium": "warning", "low": "note"}
        return mapping.get(sev.lower(), "warning")

    def to_sarif(self, report: GovernanceReport) -> LogOutput:
        """Transforms a GovernanceReport into a SARIF formatted string."""
        results_list = []

        # Convert report to internal dict representation first (logic moved from
        # adapter)
        report_data = self.report_to_dict(report)

        for adapter_name, adapter_results in report_data.items():
            if adapter_name in ["score", "is_passing", "summary"]:
                continue

            if not isinstance(adapter_results, list):
                continue

            for error in adapter_results:
                results_list.append(
                    {
                        "ruleId": f"{adapter_name}/{error.get('code', 'unknown')}",
                        "level": self._get_severity(error.get("severity", "medium")),
                        "message": {"text": error.get("message", "")},
                        "locations": [
                            {
                                "physicalLocation": {
                                    "artifactLocation": {
                                        "uri": error.get("file", "unknown")
                                    },
                                    "region": {
                                        "startLine": error.get("line", 1),
                                        "startColumn": error.get("column", 1),
                                    },
                                }
                            }
                        ],
                    }
                )

        sarif = {
            "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
            "version": "2.1.0",
            "runs": [
                {"tool": {"driver": {"name": "Auto-Linter"}}, "results": results_list}
            ],
        }
        return LogOutput(value=json.dumps(sarif, indent=2))

    def to_junit(self, report: GovernanceReport) -> LogOutput:
        """Transforms a GovernanceReport into a JUnit XML formatted string."""
        report_data = self.report_to_dict(report)
        xml = ['<?xml version="1.0" encoding="UTF-8"?>']

        total_tests = 0
        total_failures = 0
        testsuites = []

        for adapter_name, adapter_results in report_data.items():
            if adapter_name in ["score", "is_passing", "summary"]:
                continue

            if not isinstance(adapter_results, list):
                continue

            failure_count = len(adapter_results)
            testsuite_lines = []
            testsuite_lines.append(
                f'  <testsuite name="{adapter_name}" tests="{max(1, failure_count)}" failures="{failure_count}">'
            )

            if failure_count == 0:
                testsuite_lines.append(
                    f'    <testcase name="lint_{adapter_name}" classname="{adapter_name}"/>'
                )
                total_tests += 1
            else:
                for i, error in enumerate(adapter_results):
                    msg = html.escape(error.get("message", ""), quote=True)
                    testsuite_lines.append(
                        f'    <testcase name="lint_{adapter_name}_{i}" classname="{adapter_name}">'
                    )
                    testsuite_lines.append(
                        f'      <failure message="Linting failed">{msg}</failure>'
                    )
                    testsuite_lines.append("    </testcase>")
                    total_tests += 1
                    total_failures += 1

            testsuite_lines.append("  </testsuite>")
            testsuites.extend(testsuite_lines)

        xml.append(
            f'<testsuites name="Auto-Linter" tests="{total_tests}" failures="{total_failures}">'
        )
        xml.extend(testsuites)
        xml.append("</testsuites>")

        return LogOutput(value="\n".join(xml))

    def report_to_dict(self, report: GovernanceReport) -> dict[Any, Any]:
        """Converts GovernanceReport entity to a plain dictionary for formatting."""
        data = {
            "score": float(report.score),
            "is_passing": bool(report.is_passing),
            "summary": {
                "violation_count": report.violation_count,
                "adapter_count": len(report.sources),
            },
        }

        for source in report.sources:
            source_name = str(source)
            results = report.results_by_source(source)
            data[source_name] = [
                {
                    "file": str(res.file),
                    "line": int(res.line),
                    "column": int(res.column),
                    "code": str(res.code),
                    "message": str(res.message),
                    "severity": str(res.severity),
                    "enclosing_scope": str(res.enclosing_scope)
                    if res.enclosing_scope
                    else None,
                }
                for res in results
            ]
        return data

File: lint-arwaky/src-rust/capabilities/report_formatter_processor.rs
// report_formatter_processor — Capability for formatting reports (SARIF, JUnit).
// Implements ILintReportingProtocol: format, get_formatted_payload, to_sarif, to_junit.

use crate::taxonomy::*;
use serde_json::json;

/// Format constants matching Python taxonomy
pub const FORMAT_TEXT: &str = "text";
pub const FORMAT_JSON: &str = "json";
pub const FORMAT_SARIF: &str = "sarif";
pub const FORMAT_JUNIT: &str = "junit";

/// Business logic for transforming GovernanceReports into standard formats.
pub struct ReportFormatterProcessor;

impl ReportFormatterProcessor {
    pub fn new() -> Self {
        Self
    }

    /// Standard entry point for formatting as JSON/Dict.
    pub fn format(&self, report: &GovernanceReport) -> LogOutput {
        let data = self.report_to_dict(report);
        let json_str = serde_json::to_string_pretty(&data).unwrap_or_else(|_| "{}".to_string());
        LogOutput::new(json_str)
    }

    /// Unified entry point for getting a formatted payload for the Surface layer.
    pub fn get_formatted_payload(
        &self,
        report: &GovernanceReport,
        output_format: &FileFormat,
    ) -> ResponseData {
        let data = self.report_to_dict(report);

        match output_format.name.as_str() {
            "sarif" => {
                let sarif = self.to_sarif(report);
                let mut value = ResponseData::new();
                value.value = Some(json!({
                    "format": "sarif",
                    "data": sarif.value
                }));
                value
            }
            "junit" => {
                let junit = self.to_junit(report);
                let mut value = ResponseData::new();
                value.value = Some(json!({
                    "format": "junit",
                    "data": junit.value
                }));
                value
            }
            "json" => {
                let mut value = ResponseData::new();
                value.value = Some(json!({
                    "format": "json",
                    "data": data
                }));
                value
            }
            _ => {
                // Default to text/dict representation
                let mut value = ResponseData::new();
                value.value = Some(json!({
                    "format": "text",
                    "data": data
                }));
                value
            }
        }
    }

    /// Map severity string to SARIF level.
    fn _get_severity(&self, sev: &str) -> &str {
        match sev.to_lowercase().as_str() {
            "high" => "error",
            "medium" => "warning",
            "low" => "note",
            _ => "warning",
        }
    }

    /// Transform a GovernanceReport into a SARIF formatted string.
    pub fn to_sarif(&self, report: &GovernanceReport) -> LogOutput {
        let report_data = self.report_to_dict(report);
        let mut results_list: Vec<serde_json::Value> = Vec::new();

        if let Some(obj) = report_data.as_object() {
            for (adapter_name, adapter_results) in obj {
                if adapter_name == "score" || adapter_name == "is_passing" || adapter_name == "summary" {
                    continue;
                }
                if let Some(results_arr) = adapter_results.as_array() {
                    for error in results_arr {
                        let code = error.get("code").and_then(|v| v.as_str()).unwrap_or("unknown");
                        let severity = error.get("severity").and_then(|v| v.as_str()).unwrap_or("medium");
                        let message = error.get("message").and_then(|v| v.as_str()).unwrap_or("");
                        let file = error.get("file").and_then(|v| v.as_str()).unwrap_or("unknown");
                        let line = error.get("line").and_then(|v| v.as_i64()).unwrap_or(1);
                        let column = error.get("column").and_then(|v| v.as_i64()).unwrap_or(1);

                        results_list.push(json!({
                            "ruleId": format!("{}/{}", adapter_name, code),
                            "level": self._get_severity(severity),
                            "message": { "text": message },
                            "locations": [{
                                "physicalLocation": {
                                    "artifactLocation": { "uri": file },
                                    "region": {
                                        "startLine": line,
                                        "startColumn": column
                                    }
                                }
                            }]
                        }));
                    }
                }
            }
        }

        let sarif = json!({
            "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
            "version": "2.1.0",
            "runs": [{
                "tool": { "driver": { "name": "Auto-Linter" } },
                "results": results_list
            }]
        });

        LogOutput::new(serde_json::to_string_pretty(&sarif).unwrap_or_else(|_| "{}".to_string()))
    }

    /// Transform a GovernanceReport into a JUnit XML formatted string.
    pub fn to_junit(&self, report: &GovernanceReport) -> LogOutput {
        let report_data = self.report_to_dict(report);
        let mut xml_lines: Vec<String> = Vec::new();
        xml_lines.push(r#"<?xml version="1.0" encoding="UTF-8"?>"#.to_string());

        let mut total_tests: i64 = 0;
        let mut total_failures: i64 = 0;
        let mut testsuites: Vec<String> = Vec::new();

        if let Some(obj) = report_data.as_object() {
            for (adapter_name, adapter_results) in obj {
                if adapter_name == "score" || adapter_name == "is_passing" || adapter_name == "summary" {
                    continue;
                }
                if let Some(results_arr) = adapter_results.as_array() {
                    let failure_count = results_arr.len() as i64;
                    let test_count = std::cmp::max(1, failure_count);

                    testsuites.push(format!(
                        r#"  <testsuite name="{}" tests="{}" failures="{}">"#,
                        adapter_name, test_count, failure_count
                    ));

                    if failure_count == 0 {
                        testsuites.push(format!(
                            r#"    <testcase name="lint_{}" classname="{}"/>"#,
                            adapter_name, adapter_name
                        ));
                        total_tests += 1;
                    } else {
                        for (i, error) in results_arr.iter().enumerate() {
                            let msg = xml_escape(
                                error.get("message").and_then(|v| v.as_str()).unwrap_or(""),
                            );
                            testsuites.push(format!(
                                r#"    <testcase name="lint_{}_{}" classname="{}">"#,
                                adapter_name, i, adapter_name
                            ));
                            testsuites.push(format!(
                                r#"      <failure message="Linting failed">{}</failure>"#,
                                msg
                            ));
                            testsuites.push("    </testcase>".to_string());
                            total_tests += 1;
                            total_failures += 1;
                        }
                    }

                    testsuites.push("  </testsuite>".to_string());
                }
            }
        }

        xml_lines.push(format!(
            r#"<testsuites name="Auto-Linter" tests="{}" failures="{}">"#,
            total_tests, total_failures
        ));
        xml_lines.extend(testsuites);
        xml_lines.push("</testsuites>".to_string());

        LogOutput::new(xml_lines.join("\n"))
    }

    /// Converts GovernanceReport entity to a plain dictionary for formatting.
    pub fn report_to_dict(&self, report: &GovernanceReport) -> serde_json::Value {
        let violation_count = report.violation_count().value;
        let adapter_count = report.results.values.iter()
            .map(|r| r.source.as_ref().map(|s| s.value.clone()).unwrap_or_default())
            .collect::<std::collections::HashSet<_>>()
            .len();

        let mut data = serde_json::json!({
            "score": report.score.value,
            "is_passing": report.is_passing.value,
            "summary": {
                "violation_count": violation_count,
                "adapter_count": adapter_count
            }
        });

        // Group results by source adapter
        let mut by_source: std::collections::HashMap<String, Vec<serde_json::Value>> =
            std::collections::HashMap::new();

        for result in &report.results.values {
            let source_name = result.source.as_ref()
                .map(|s| s.value.clone())
                .unwrap_or_else(|| "unknown".to_string());

            let mut entry = serde_json::json!({
                "file": result.file.to_string(),
                "line": result.line.value,
                "column": result.column.value,
                "code": result.code.to_string(),
                "message": result.message.to_string(),
                "severity": result.severity.to_string(),
                "enclosing_scope": result.enclosing_scope.as_ref().map(|s| s.to_string())
            });

            // Remove null enclosing_scope for cleaner output
            if entry.get("enclosing_scope").and_then(|v| v.as_null()).is_some() {
                entry.as_object_mut().map(|m| m.remove("enclosing_scope"));
            }

            by_source.entry(source_name).or_default().push(entry);
        }

        if let Some(obj) = data.as_object_mut() {
            for (source_name, results) in by_source {
                obj.insert(source_name, serde_json::Value::Array(results));
            }
        }

        data
    }
}

/// Escape special XML characters.
fn xml_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&apos;")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_severity() {
        let formatter = ReportFormatterProcessor::new();
        assert_eq!(formatter._get_severity("high"), "error");
        assert_eq!(formatter._get_severity("medium"), "warning");
        assert_eq!(formatter._get_severity("low"), "note");
        assert_eq!(formatter._get_severity("unknown"), "warning");
    }

    #[test]
    fn test_xml_escape() {
        assert_eq!(xml_escape("a < b & c > d"), "a &lt; b &amp; c &gt; d");
        assert_eq!(xml_escape(r#"he said "hi""#), "he said &quot;hi&quot;");
    }

    #[test]
    fn test_report_to_dict() {
        let formatter = ReportFormatterProcessor::new();
        let report = GovernanceReport::new();
        let dict = formatter.report_to_dict(&report);
        assert_eq!(dict["score"], 100.0);
        assert_eq!(dict["is_passing"], true);
    }
}
